### PR TITLE
sockets: Improve behavior when sending to closed connection

### PR DESCRIPTION
**Pull request by comex**
[This is based on my SSL PR as otherwise there is a merge conflict.  However, it doesn't seem that I can set the base of this PR to that one while still having this count as a pull request on yuzu-emu/yuzu repo. 
 So I guess, only look at the [bottom-most commit](https://github.com/yuzu-emu/yuzu/pull/10989/commits/7949bcd151760320ad9c6154ded40e62d6acf892)?]

- On Unix, this would previously kill the Yuzu process with SIGPIPE.  Send MSG_NOSIGNAL to opt out of this.

- Add support for the proper error code in this situation, EPIPE.

- Windows has nonstandard behavior in this situation; translate it to the standard behavior.  Kind of pointless, but isn't it nice to be correct?

### CODE REVIEW

**Comment by liamwhite** [comment]
Needs a rebase and CI fixes

File: src/common/socket_types.h [hunk 1]
```diff
@@ -5,15 +5,19 @@
...
 #include "common/common_types.h"
 
+#include <optional>
+
 namespace Network {
 
 /// Address families
 enum class Domain : u8 {
-    INET, ///< Address family for IPv4
+    Unspecified, ///< Represents 0, used in getaddrinfo hints
```
Comment (Tachi107) [inline src/common/socket_types.h:14]: Shouldn't this be `UNSPEC` for consistency with `AF_UNSPEC`?

File: src/core/hle/service/sockets/sfdnsres.cpp [hunk 2]
```diff
@@ -59,188 +50,285 @@ enum class NetDbError : s32 {
...
+    Append<u16_be>(data, static_cast<u16>(Domain::INET)); // h_addrtype
+    Append<u16_be>(data, sizeof(Network::IPv4Address));   // h_length
+    // h_addr_list:
+    size_t count = vec.size();
+    ASSERT(count <= UINT32_MAX);
+    Append<u32_be>(data, static_cast<uint32_t>(count));
+    for (const Network::AddrInfo& addrinfo : vec) {
+        // On the Switch, this is passed through htonl despite already being
+        // big-endian, so it ends up as little-endian.
+        Append<u32_le>(data, Network::IPv4AddressToInteger(addrinfo.addr.ip));
```
Comment (Tachi107) [inline src/core/hle/service/sockets/sfdnsres.cpp:125]: Are you sure? `htonl()` is a no-op on big-endian machines.

Also, mostly a curiosity: I see that a lot of BSD sockets code in yuzu assumes IPv4 addresses, is the Switch IPv4-only?

File: src/core/hle/service/ssl/ssl.cpp [hunk 3]
```diff
@@ -80,21 +107,299 @@ class ISslConnection final : public ServiceFramework<ISslConnection> {
...
+            u32 offset;
+        };
+        if (!get_server_cert_chain) {
+            // Just return the first one, unencoded.
+            ASSERT_OR_EXECUTE_MSG(
+                !certs.empty(), { return {}; }, "Should be at least one server cert");
+            return certs[0];
+        }
+        std::vector<u8> ret;
+        Header header{0x4E4D684374726543, static_cast<u32>(certs.size()), 0};
```
Comment (Tachi107) [inline src/core/hle/service/ssl/ssl.cpp:235]: A commend about the meaning of this number would be helpful

File: src/core/internal_network/network.cpp [hunk 4]
```diff
@@ -219,30 +243,101 @@ Errno TranslateNativeError(int e) {
...
+            return GetAddrInfoError::NODATA;
+        }
+#endif
+        return GetAddrInfoError::OTHER;
+    }
+}
+
+Domain TranslateDomainFromNative(int domain) {
     switch (domain) {
+    case 0:
```
Comment (Tachi107) [inline src/core/internal_network/network.cpp:327]: Should this be `AF_UNSPEC`?

**Review by Tachi107 — COMMENTED** [review]
Just some feedback/questions

File: src/core/hle/service/sockets/sfdnsres.cpp (hunk 2 above)
Comment (liamwhite) [reply src/core/hle/service/sockets/sfdnsres.cpp:125]: The switch has no support at all for ipv6.

File: src/core/hle/service/ssl/ssl.cpp (hunk 3 above)
Comment (liamwhite) [reply src/core/hle/service/ssl/ssl.cpp:235]: "CertChMN"

**Review by liamwhite — APPROVED** [review]

**Comment by Tachi107** [comment]
Comment https://github.com/yuzu-emu/yuzu/pull/10989#discussion_r1278051635 remains unanswered :/

**Pull request merged**